src/dbus_api.rs

```diff
@@ -141,10 +143,111 @@ fn list_pools(m: &MethodInfo<MTFn<TData>, TData>) -> MethodResult {
     Ok(vec![msg])
 }
 
+fn create_dbus_filesystem<'a>(mut dbus_context: DbusContext) -> dbus::Path<'a> {
+
+    let f = Factory::new_fn();
+
+    let create_snapshot_method = f.method(CREATE_SNAPSHOT, (), create_filesystems)
+        .in_arg(("name", "s"))
+        .out_arg(("object_path", "o"))
+        .out_arg(("return_code", "q"))
+        .out_arg(("return_string", "s"));
+
+    let rename_method = f.method(RENAME, (), destroy_filesystems)
+        .in_arg(("name", "s"))
+        .out_arg(("object_path", "o"))
+        .out_arg(("return_code", "q"))
+        .out_arg(("return_string", "s"));
+
+    let set_mountpoint_method = f.method(SET_MOUNTPOINT, (), list_filesystems)
+        .out_arg(("object_path", "o"))
+        .out_arg(("return_code", "q"))
+        .out_arg(("return_string", "s"));
+
+    let set_quota_method = f.method(SET_QUOTA, (), list_cache_devs)
+        .out_arg(("quota", "s"))
+        .out_arg(("object_path", "o"))
+        .out_arg(("return_code", "q"))
+        .out_arg(("return_string", "s"));
+
+
+    let object_name = format!("{}/{}",
+                              STRATIS_BASE_PATH,
+                              dbus_context.get_next_id().to_string());
+
+    let object_path = f.object_path(object_name, dbus_context.clone())
+        .introspectable()
+        .add(f.interface(STRATIS_FILESYSTEM_BASE_INTERFACE, ())
+            .add_m(create_snapshot_method)
+            .add_m(rename_method)
+            .add_m(set_mountpoint_method)
+            .add_m(set_quota_method));
+
+    let path = object_path.get_name().to_owned();
+    dbus_context.action_list.borrow_mut().push(DeferredAction::Add(object_path));
+    path
+}
 
 fn create_filesystems(m: &MethodInfo<MTFn<TData>, TData>) -> MethodResult {
+    let message: &Message = m.msg;
+    let mut iter = message.iter_init();
 
-    Ok(vec![m.msg.method_return().append3("/dbus/cache/path", 0, "Ok")])
+    if iter.arg_type() == 0 {
+        return Err(MethodErr::no_arg());
+    }
+
+    let mut filesystems: Array<(&str, &str, &str), _> =
+        try!(iter.read::<Array<(&str, &str, &str), _>>()
+            .map_err(|_| MethodErr::invalid_arg(&0)));
+
+    let dbus_context = m.path.get_data();
+    let object_path = m.path.get_name();
+    let return_message = message.method_return();
+
+    println!("Object Path {}", object_path);
+
+    let pool_name = match dbus_context.pools.borrow_mut().get(&object_path.to_string()) {
+        Some(pool) => pool.clone(),
+        None => return Err(MethodErr::invalid_arg(&0)),
+    };
+
+    let mut b_engine = dbus_context.engine.borrow_mut();
+    let ref mut pool = match b_engine.get_pool(&pool_name) {
+        Ok(result) => result,
+        Err(x) => {
+            let (rc, rs) = engine_to_dbus_err(&x);
+            let (rc, rs) = code_to_message_items(rc, rs);
+            let entry = MessageItem::Struct(vec!["o".into(), "q".into(), "s".into()]);
```

Comment:
I think it is necessary to use specific correct MessageItem constructors for the components of the message.